File: java-app/src/main/java/com/dexter/activities/crop/CropImageActivity.java
package com.dexter.activities.crop;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.Toast;

import com.dexter.utils.Constants;
import com.dexter.utils.R;
import com.yalantis.ucrop.UCrop;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Date;

import timber.log.Timber;

public class CropImageActivity extends AppCompatActivity {

    private static final String TAG = "CropImageActivity";
    private static final String IMG_PREFIX = "IMG_";
    private static final String IMG_SUFFIX = ".jpg";
    private float x;
    private float y;

    public static Intent getStartIntent(Context context, Uri data, float x, float y) {
        Intent intent = new Intent(context, CropImageActivity.class);
        intent.putExtra(Constants.uri, data);
        intent.putExtra(Constants.x, x);
        intent.putExtra(Constants.y, y);
        return intent;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent intent = getIntent();
        Uri uri = intent.getParcelableExtra(Constants.uri);
        x = intent.getFloatExtra(Constants.x, -1);
        y = intent.getFloatExtra(Constants.y, -1);
        if (uri != null) {
            startCropActivity(uri);
        } else {
            finish();
        }
    }

    private void startCropActivity(@NonNull Uri uri) {
        @SuppressLint("SimpleDateFormat")
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
        String path = IMG_PREFIX + timeStamp + IMG_SUFFIX;
        UCrop uCrop = UCrop.of(uri, Uri.fromFile(new File(getCacheDir(), path)));
        uCrop = advancedConfig(uCrop);
        uCrop.start(CropImageActivity.this);
    }

    /**
     * Sometimes you want to adjust more options, it's done via {@link com.yalantis.ucrop.UCrop.Options} class.
     *
     * @param uCrop - ucrop builder instance
     * @return - ucrop builder instance
     */
    private UCrop advancedConfig(@NonNull UCrop uCrop) {
        UCrop.Options options = new UCrop.Options();
        options.withAspectRatio(x, y);
        return uCrop.withOptions(options);
    }

    private void handleCropResult(@NonNull Intent result) {
        final Uri resultUri = UCrop.getOutput(result);
        if (resultUri != null) {
            Intent intent = new Intent();
            intent.setData(resultUri);
            setResult(RESULT_OK, intent);
            finish();
        } else {
            Toast.makeText(CropImageActivity.this, R.string.toast_cannot_retrieve_cropped_image, Toast.LENGTH_SHORT).show();
        }
    }

    @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
    private void handleCropError(@NonNull Intent result) {
        final Throwable cropError = UCrop.getError(result);
        if (cropError != null) {
            Timber.d(TAG);
            Timber.e("handleCropError: " + cropError);
            Toast.makeText(CropImageActivity.this, cropError.getMessage(), Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(CropImageActivity.this, R.string.toast_unexpected_error, Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            if (requestCode == UCrop.REQUEST_CROP) {
                handleCropResult(data);
            }
        }
        if (resultCode == UCrop.RESULT_ERROR) {
            handleCropError(data);
        } else {
            finish();
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }
}

File: kotlin-app/src/main/java/com/dexter/activities/crop/CropImageActivity.kt
package com.dexter.activities.crop

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.Toast
import com.dexter.utils.Constants
import com.dexter.utils.R
import com.yalantis.ucrop.UCrop
import timber.log.Timber
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class CropImageActivity : AppCompatActivity() {
    private var x: Float = 0.toFloat()
    private var y: Float = 0.toFloat()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = intent
        val uri = intent.getParcelableExtra<Uri>(Constants.uri)
        x = intent.getFloatExtra(Constants.x, -1f)
        y = intent.getFloatExtra(Constants.y, -1f)
        if (uri != null) {
            startCropActivity(uri)
        } else {
            finish()
        }
    }

    private fun startCropActivity(uri: Uri) {
        @SuppressLint("SimpleDateFormat")
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        val path = IMG_PREFIX + timeStamp + IMG_SUFFIX
        var uCrop = UCrop.of(uri, Uri.fromFile(File(cacheDir, path)))
        uCrop = advancedConfig(uCrop)
        uCrop.start(this@CropImageActivity)
    }

    /**
     * Sometimes you want to adjust more options, it's done via [UCrop.Options] class.
     *
     * @param uCrop - ucrop builder instance
     * @return - ucrop builder instance
     */
    private fun advancedConfig(uCrop: UCrop): UCrop {
        val options = UCrop.Options()
        options.withAspectRatio(x, y)
        return uCrop.withOptions(options)
    }

    private fun handleCropResult(result: Intent) {
        val resultUri = UCrop.getOutput(result)
        if (resultUri != null) {
            val intent = Intent()
            intent.data = resultUri
            setResult(Activity.RESULT_OK, intent)
            finish()
        } else {
            Toast.makeText(this@CropImageActivity, R.string.toast_cannot_retrieve_cropped_image, Toast.LENGTH_SHORT).show()
        }
    }

    private fun handleCropError(result: Intent) {
        val cropError = UCrop.getError(result)
        if (cropError != null) {
            Timber.d(TAG)
            Timber.e("handleCropError: $cropError")
            Toast.makeText(this@CropImageActivity, cropError.message, Toast.LENGTH_LONG).show()
        } else {
            Toast.makeText(this@CropImageActivity, R.string.toast_unexpected_error, Toast.LENGTH_SHORT).show()
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == UCrop.REQUEST_CROP) {
                handleCropResult(data!!)
            }
        }
        if (resultCode == UCrop.RESULT_ERROR) {
            handleCropError(data!!)
        } else {
            finish()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
    }

    companion object {

        private const val TAG = "CropImageActivity"
        private const val IMG_PREFIX = "IMG_"
        private const val IMG_SUFFIX = ".jpg"

        fun getStartIntent(context: Context, data: Uri, x: Float, y: Float): Intent {
            val intent = Intent(context, CropImageActivity::class.java)
            intent.putExtra(Constants.uri, data)
            intent.putExtra(Constants.x, x)
            intent.putExtra(Constants.y, y)
            return intent
        }
    }
}
